README
Unix-Statgrab version 0.04
==========================

Unix::Statgrab is a wrapper for libstatgrab, as available from
http://www.i-scream.org/libstatgrab/. It is a reasonably portable 
attempt to query interesting stats about your computer. It covers 
information on the operating system, CPU, memory usage, network 
interfaces, hard-disks etc.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

    libstatgrab from http://www.i-scream.org/libstatgrab/

Version 0.12 of libstatgrab is required.

If you have installed libstatgrab in a non-standard location,
you have to tell Makefile.PL about it:

    perl Makefile.PL LIBS="-L/path/to/location -lstatgrab" INC="-I/path/to/includes"

COPYRIGHT AND LICENCE

Copyright (C) 2004-2005 by Tassilo von Parseval

This library is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 2.1 of the License, or (at your option) any
later version.



Changes
Revision history for Perl extension Unix::Statgrab.

0.110_003	2016-04-21
    - add missing prove whether requested row exists in stat or not for
      fetchrow_hashref and fetchrow_arrayref calls

0.110_002	2016-04-20
    - harden passing strings from libstatgrab to perl when fetching stats
      as array

0.110_001	2016-04-14
    - harden tests
      * eval around all statgrab calls & don't prove result when exception catched
      * check for result and error, if no result received
    - harden passing strings from libstatgrab to perl when fetching stats
      as hash
    - improve author tests and so improve documentation

0.109	2015-07-10
    - documentation improvements using pod and spelling checker
    - remove upstream included "_default_headers_with_perl" C::AC method

0.108_001	2015-02-03
    - add full flavored struct accessors and implement all in XS, supports now
      colnames, fetchrow_arrayref, fetchall_arrayref, fetchrow_hashref, fetchall_hashref,
      fetchall_hash fetchall_array fetchall_table
    - as_list is deprecated in favor of fetchall_hash

0.107	2015-01-20
    - eliminate potential memory leaks in get_fs_stats, get_network_io_stats,
      get_network_iface_stats, get_process_stats
    - add support for device_canonical
    - update bundled ppport.h to 3.28 one

0.016	2015-01-06
    - update meta-information
    - update author tests
    - take advantage of Config::AutoConf 0.306

0.105_001 2014-09-07
    - take advantage of Config::AutoConf 0.23
    - switch to ExtUtils::MakeMaker, it seems more active maintained
    - add a test for possible leak
    - improve repository documentation (README.md)

0.104 2014-02-20
    - fix link targets for stats description

0.103 2014-02-19
    - learn include paths and linker flags from pkg-config, if available
    - keep configure phase settings persistent for building stage
    - redo meta_add using meta-spec v2
      * includes developer deps to please ribasushi

0.102 2013-09-28
    - Update repository URI - it's now officially hosted on i-scream.org
      with a mirror on GitHub for issue tracking and easy deal with
      pull requests

0.101 2013-09-10
    - depend on 'parent' not 'parant' (Thanks to Chris Weyl)

0.100 2013-09-09
    - add POD for generic methods
    - return stats in integer representation, if valid (read: if
      UV can hold any value of unsigned long long)
    - add "parants" to configure_requires

0.099_002 2013-09-06
    - Update POD to match current functionality
    - introduce as_list() for each object
    - Changes reformatted as per CPAN::Changes::Spec
    - Fix Build to make the distribution more CPAN testers friendly

0.099_001 2013-08-03 19:13:04 CEST
    - Update Unix::Statgrab for being used with libstatgrab 0.90
    - kick out all "special" handling (process_stats, cpu_percents)
      and replace them by general stats as naturally provided
    - move pod tests to xt/ making them Author only tests

0.04  2005-09-22 08:20:59 CEST
    - 0.12 is now required for compilation
    - therefore: additional statistics for get_fs_stats()
    - fixed a minor nit in the POD tests:
      make sure that Test::Pod and Test::Pod::Coverage are
      installed in the required minimum versions
      (thanks to Ricardo Signes <rjbs AT cpan DOT org>)

0.03  2005-04-03 07:38:20 CEST
    - POD- and POD-coverage-tests

0.02  2004-06-21 07:05:16 CEST
    - running Makefile.PL could nuke /dev/null. Fixed.
      (Tim Bishop <tim AT bishnet DOT net>)
    - some DESTROY methods missing which potentially lead
      to some slight memory leaks. Fixed.
    - added sg_get_process_stats(). I must have missed that one
      for some reason.

0.01  2004-05-21 06:57:16
    - original version; created by h2xs 1.23 with
      "h2xs -O -b 5.5.3 -n Unix::Statgrab /usr/local/include/statgrab.h"



Build.PL
use 5.008003;

use strict;
use warnings;

use lib qw(inc);
use Statgrab::Builder;

my $builder = Statgrab::Builder->new(
                          module_name => 'Unix::Statgrab',
                          license     => 'perl',
                          dist_author => [
                                       'Tassilo von Parseval <tassilo.von.parseval@rwth-aachen.de>',
                                       'Jens Rehsack <sno@NetBSD.org>'
                          ],
                          needs_compiler     => 1,
                          configure_requires => {
                                                  'File::Copy'       => '0',
                                                  'Module::Build'    => '0.4000',
                                                  'Config::AutoConf' => '0.23',
                                                  'parent'           => 0,
                                                },
                          build_requires => {
                                              'Config::AutoConf'   => '0.19',
                                              'Capture::Tiny'      => '0.06',
                                              'ExtUtils::CBuilder' => '0.27',
                                              'Module::Build'      => '0.4000',
                                            },
                          meta_merge => {
                                         'meta-spec' => { version => 2 },
                                         resources   => {
                                             repository => {
                                                 web => 'https://github.com/i-scream/Unix-Statgrab',
                                                 type => 'git',
                                             },
                                         },
                                         prereqs => {
                                                 develop => {
                                                     requires => {
                                                         'Test::CPAN::Changes'                 => 0,
                                                         'Test::CheckManifest'                 => 0,
                                                         'Test::Pod'                           => 0,
                                                         'Test::Pod::Coverage'                 => 0,
                                                         'Test::Pod::Spelling::CommonMistakes' => 0,
                                                     },
                                                 },
                                         },
                                        },
                          add_to_cleanup => [ 'Unix-Statgrab-*', '*.o', 'config.h', 'config.log', ],
                          test_files => [ "t/*.t", "xt/*.t" ],
);
eval { $builder->dispatch("configure"); };
$@ and warn $@ and exit 0;
$builder->create_build_script();


inc/Statgrab/Builder.pm
package Statgrab::Builder;

use warnings;
use strict;

use parent 'Module::Build';

use Config;
use Carp;

use Config::AutoConf::SG ();

use Cwd qw(getcwd);
use File::Spec ();

use AutoSplit             ();
use ExtUtils::Constant    ();
use ExtUtils::Mkbootstrap ();
use ExtUtils::ParseXS     ();

sub ACTION_configure
{
    my $self = shift;

    my $autoconf = Config::AutoConf::SG->new();
    $autoconf->check_libstatgrab() or die <<EOD;
*******************************************
Couldn't find libstatgrab (at least 0.90)
which is required for this module.

To obtain it, go to
    http://www.i-scream.org/libstatgrab/
*******************************************
EOD

    $autoconf->check_sizeof_IVUV_fit_stat();

    my $xsfile   = "lib/Unix/Statgrab.xs";
    my $spec     = $self->_infer_xs_spec($xsfile);
    my $config_h = File::Spec->catfile( $spec->{src_dir}, 'config.h' );
    $autoconf->write_config_h($config_h);
    $self->config_data( extra_preprocess_flags => $autoconf->{extra_preprocess_flags} );
    $self->config_data( extra_link_flags    => $autoconf->{extra_link_flags} );

    return;
}

sub ACTION_write_constants
{
    my $self = shift;

    my @names = (
        qw(SG_ERROR_NONE SG_ERROR_INVALID_ARGUMENT SG_ERROR_ASPRINTF
          SG_ERROR_SPRINTF SG_ERROR_DEVICES SG_ERROR_DEVSTAT_GETDEVS
          SG_ERROR_DEVSTAT_SELECTDEVS SG_ERROR_DISKINFO SG_ERROR_ENOENT
          SG_ERROR_GETIFADDRS SG_ERROR_GETMNTINFO SG_ERROR_GETPAGESIZE
          SG_ERROR_HOST SG_ERROR_KSTAT_DATA_LOOKUP SG_ERROR_KSTAT_LOOKUP
          SG_ERROR_KSTAT_OPEN SG_ERROR_KSTAT_READ SG_ERROR_KVM_GETSWAPINFO
          SG_ERROR_KVM_OPENFILES SG_ERROR_MALLOC SG_ERROR_MEMSTATUS
          SG_ERROR_OPEN SG_ERROR_OPENDIR SG_ERROR_READDIR SG_ERROR_PARSE
          SG_ERROR_PDHADD SG_ERROR_PDHCOLLECT SG_ERROR_PDHOPEN SG_ERROR_PDHREAD
          SG_ERROR_PERMISSION SG_ERROR_PSTAT SG_ERROR_SETEGID SG_ERROR_SETEUID
          SG_ERROR_SETMNTENT SG_ERROR_SOCKET SG_ERROR_SWAPCTL SG_ERROR_SYSCONF
          SG_ERROR_SYSCTL SG_ERROR_SYSCTLBYNAME SG_ERROR_SYSCTLNAMETOMIB
          SG_ERROR_SYSINFO SG_ERROR_MACHCALL SG_ERROR_IOKIT SG_ERROR_UNAME
          SG_ERROR_UNSUPPORTED SG_ERROR_XSW_VER_MISMATCH SG_ERROR_GETMSG
          SG_ERROR_PUTMSG SG_ERROR_INITIALISATION SG_ERROR_MUTEX_LOCK
          SG_ERROR_MUTEX_UNLOCK),
        qw(sg_unknown_configuration sg_physical_host sg_virtual_machine
          sg_paravirtual_machine sg_hardware_virtualized),
        qw(sg_fs_unknown sg_fs_regular sg_fs_special sg_fs_loopback
          sg_fs_remote sg_fs_local sg_fs_alltypes),
        qw(SG_IFACE_DUPLEX_FULL SG_IFACE_DUPLEX_HALF SG_IFACE_DUPLEX_UNKNOWN),
        qw(SG_IFACE_DOWN SG_IFACE_UP),
        qw(SG_PROCESS_STATE_RUNNING SG_PROCESS_STATE_SLEEPING
          SG_PROCESS_STATE_STOPPED SG_PROCESS_STATE_ZOMBIE
          SG_PROCESS_STATE_UNKNOWN),
    );

    my $xsfile  = "lib/Unix/Statgrab.xs";
    my $spec    = $self->_infer_xs_spec($xsfile);
    my $c_const = File::Spec->catfile( $spec->{src_dir}, 'const-c.inc' );
    #my $c_const = File::Spec->catfile(getcwd(), 'const-c.inc');
    my $xs_const = File::Spec->catfile( $spec->{src_dir}, 'const-xs.inc' );
    #my $xs_const = File::Spec->catfile(getcwd(), 'const-xs.inc');

    # workaround #ifdef around each constant (stupid that it's not documented)
    open( my $c_fh, ">", $c_const ) or die("Cannot open $c_const for writing: $!");
    foreach my $name (@names)
    {
        print $c_fh "#define $name $name\n";
    }
    ExtUtils::Constant::WriteConstants(
                                        NAME         => 'Unix::Statgrab',
                                        NAMES        => \@names,
                                        DEFAULT_TYPE => 'UV',
                                        C_FH         => $c_fh,
                                        XS_FILE      => $xs_const,
                                      );
    close($c_fh);
    $self->add_to_cleanup($c_const);     ## FIXME
    $self->add_to_cleanup($xs_const);    ## FIXME

    return;
}

sub ACTION_autosplit
{
    my $self = shift;
    AutoSplit::autosplit( "blib/lib/Unix/Statgrab.pm", "blib/arch/auto", 0, 1, 1 );
}

sub ACTION_code
{
    my $self     = shift;
    my $cbuilder = $self->cbuilder;

    $self->{properties}->{extra_linker_flags}   ||= [];
    $self->{properties}->{extra_compiler_flags} ||= [];
    unshift(
             @{ $self->{properties}->{extra_linker_flags} },
             split( " ", $cbuilder->{config}{ldflags} )
           );    # XXX currently silently ignored by EU::CB
    push(
          @{ $self->{properties}->{extra_linker_flags} },
          @{ $self->config_data("extra_link_flags") }
        );
    push(
          @{ $self->{properties}->{extra_compiler_flags} },
          @{ $self->config_data("extra_preprocess_flags") }
        );

    # for my $path (catdir("blib","bindoc"), catdir("blib","bin")) {
    #     mkpath $path unless -d $path;
    # }

    #$self->dispatch("configure");
    $self->dispatch("write_constants");

    $self->SUPER::ACTION_code();

    $self->dispatch("autosplit");
}

1;


inc/Config/AutoConf/SG.pm
package Config::AutoConf::SG;

use strict;
use warnings;

use parent qw(Config::AutoConf);
use Capture::Tiny qw/capture/;
use Text::ParseWords qw(shellwords);

sub new
{
    my ( $class, %args ) = @_;
    my $self = $class->SUPER::new(%args);
    # XXX might add c++ if required for some operating systems
    return $self;
}

sub check_libstatgrab
{
    my ($self) = @_;
    ref($self) or $self = $self->_get_instance();
    $self->pkg_config_package_flags( "libstatgrab" );

    return $self->search_libs( "sg_get_process_stats_r", ["statgrab"] );
}

sub check_device_canonical
{
    # device_canonical
    my ($self) = @_;
    ref($self) or $self = $self->_get_instance();

    $self->check_member( "sg_fs_stats.device_canonical", { prologue => join("\n", $self->_default_includes, "#include <statgrab.h>") } );
}

my %TYPE_ASSO = (
                  UV => "unsigned long long",
                  IV => "long long"
                );

sub _check_sizeof_type_fit_stat
{
    my ( $self, $type ) = @_;

    my $includes    = $self->_default_includes_with_perl();
    my $match_check = "sizeof($type) == sizeof($TYPE_ASSO{$type})";
    my $conftest    = $self->lang_build_bool_test( $includes, $match_check );

    my $match_size = $self->compile_if_else($conftest);

    return $match_size;
}

sub _match_type_define_name
{
    my $type            = $_[0];
    my $match_name      = "MATCH_" . uc($type) . "_FIT_";
    my @type_asso_words = split( " ", $TYPE_ASSO{$type} );
    foreach my $taw (@type_asso_words) { $match_name .= uc( substr( $taw, 0, 1 ) ) }
    return $match_name;
}

sub check_sizeof_IVUV_fit_stat
{
    my ($self) = @_;

    my $match = 1;
    foreach my $type (qw(IV UV))
    {
        my $cache_size_match_name = $self->_cache_name( "type", $type, qw(size match ll) );
        my $check_size_match_sub = sub {
            my $match_size = $self->_check_sizeof_type_fit_stat($type);
            $self->define_var( _match_type_define_name($type),
                               $match_size ? $match_size : undef,
                               "defined when $type can hold any $TYPE_ASSO{$type} value" );

            return $match_size;
        };

        $match &=
          $self->check_cached( $cache_size_match_name,
                               "if $type can hold any $TYPE_ASSO{$type} value",
                               $check_size_match_sub );
    }

    return $match;
}

1;


Makefile.PL
#!perl

use strict;
use warnings;

use lib 'inc';
use Config;
use Config::AutoConf::SG;
use ExtUtils::Constant;
use ExtUtils::MakeMaker;
use File::Basename;
use File::Spec;
use Text::ParseWords;

my %RUN_DEPS = (
    'Carp'           => 0,
    'File::Basename' => 0,
    'File::Path'     => '2.00',
    'File::Spec'     => 0,
);
my %CONFIGURE_DEPS = (
    'ExtUtils::MakeMaker' => 0,
    'ExtUtils::Constant'  => 0,
    'Capture::Tiny'       => '0.06',
    'ExtUtils::CBuilder'  => '0.27',
    'Config::AutoConf'    => '0.308',
    'File::Basename'      => 0,
    'File::Spec'          => 0,
    'Text::ParseWords'    => 0,
    'parent'              => 0,
);
my %BUILD_DEPS = ();

my %TEST_DEPS = (
    'Test::More' => 0.90,
);

my $autoconf = Config::AutoConf::SG->new();
$autoconf->check_libstatgrab() or die <<EOD;
*******************************************
Couldn't find libstatgrab (at least 0.90)
which is required for this module.

To obtain it, go to
    http://www.i-scream.org/libstatgrab/
*******************************************
EOD

$autoconf->check_device_canonical();
$autoconf->check_sizeof_IVUV_fit_stat();

my $xsfile = "Statgrab.xs";
my $config_h = File::Spec->catfile( dirname($xsfile), 'config.h' );
$autoconf->write_config_h($config_h);
_write_constants();

WriteMakefile1(
    MIN_PERL_VERSION => '5.008003',
    META_ADD         => {
        'meta-spec' => { version => 2 },
        resources   => {
            homepage   => 'https://metacpan.org/release/Unix-Statgrab',
            repository => {
                url  => 'https://github.com/i-scream/Unix-Statgrab.git',
                web  => 'https://github.com/i-scream/Unix-Statgrab',
                type => 'git',
            },
            bugtracker => {
                web    => 'http://rt.cpan.org/Public/Dist/Display.html?Name=Unix-Statgrab',
                mailto => 'bug-Unix-Statgrab@rt.cpan.org',
            },
            license       => 'http://dev.perl.org/licenses/',
            x_IRC         => "irc://irc.freenode.net/#libstatgrab",
            x_MailingList => "mailto:users\@i-scream.org'",
        },
        prereqs => {
            develop => {
                requires => {
                    'Test::CPAN::Changes'                 => 0,
                    'Test::CheckManifest'                 => 0,
                    'Module::CPANTS::Analyse'             => '0.96',
                    'Test::Kwalitee'                      => 0,
                    'Test::Pod'                           => 0,
                    'Test::Pod::Coverage'                 => 0,
                    'Test::Pod::Spelling::CommonMistakes' => 0,
                    'Test::Spelling'                      => 0,
                },
            },
            configure => {
                requires => {%CONFIGURE_DEPS},
            },
            build => { requires => {%BUILD_DEPS} },
            test  => {
                recommends => { 'Test::LeakTrace' => 0 },
                requires   => {%TEST_DEPS}
            },
            runtime => {
                requires => { %RUN_DEPS },
            },
        },
    },
    NAME               => 'Unix::Statgrab',
    VERSION_FROM       => 'lib/Unix/Statgrab.pm',
    ABSTRACT_FROM      => 'lib/Unix/Statgrab.pm',
    LICENSE            => 'perl',
    AUTHOR             => [ 'Tassilo von Parseval <tassilo.von.parseval@rwth-aachen.de>', 'Jens Rehsack <sno@NetBSD.org>' ],
    XS                 => { $xsfile => "Statgrab.c" },
    OBJECT             => q/$(O_FILES)/,
    INC                => join( " ", @{ $autoconf->{extra_preprocess_flags} } ),
    EXTRALIBS          => join( " ", @{ $autoconf->{extra_link_flags} } ),
    LDLOADLIBS         => join( " ", @{ $autoconf->{extra_link_flags} } ),
    CONFIGURE_REQUIRES => \%CONFIGURE_DEPS,
    PREREQ_PM          => \%RUN_DEPS,
    BUILD_REQUIRES     => \%BUILD_DEPS,
    TEST_REQUIRES      => \%TEST_DEPS,
    clean              => { FILES => "Unix-Statgrab-* *.o config.h const-*.inc config.log" },
    test               => { TESTS => 't/*.t xt/*.t' },
);

sub WriteMakefile1
{    # originally written by Alexandr Ciornii, version 0.21. Added by eumm-upgrade.
    my %params       = @_;
    my $eumm_version = $ExtUtils::MakeMaker::VERSION;
    $eumm_version = eval $eumm_version;
    die "EXTRA_META is deprecated" if ( exists( $params{EXTRA_META} ) );
    die "License not specified"    if ( !exists( $params{LICENSE} ) );
    $params{TEST_REQUIRES}
      and $eumm_version < 6.6303
      and $params{BUILD_REQUIRES} = { %{ $params{BUILD_REQUIRES} || {} }, %{ delete $params{TEST_REQUIRES} } };
    #EUMM 6.5502 has problems with BUILD_REQUIRES
    $params{BUILD_REQUIRES}
      and $eumm_version < 6.5503
      and $params{PREREQ_PM} = { %{ $params{PREREQ_PM} || {} }, %{ delete $params{BUILD_REQUIRES} } };
    ref $params{AUTHOR}
      and "ARRAY" eq ref $params{AUTHOR}
      and $eumm_version < 6.5702
      and $params{AUTHOR} = join( ", ", @{ $params{AUTHOR} } );
    delete $params{CONFIGURE_REQUIRES}    if ( $eumm_version < 6.52 );
    delete $params{MIN_PERL_VERSION}      if ( $eumm_version < 6.48 );
    delete $params{META_MERGE}            if ( $eumm_version < 6.46 );
    delete $params{META_ADD}{prereqs}     if ( $eumm_version < 6.58 );
    delete $params{META_ADD}{'meta-spec'} if ( $eumm_version < 6.58 );
    delete $params{META_ADD}              if ( $eumm_version < 6.46 );
    delete $params{LICENSE}               if ( $eumm_version < 6.31 );
    delete $params{AUTHOR}                if ( $] < 5.005 );
    delete $params{ABSTRACT_FROM}         if ( $] < 5.005 );
    delete $params{BINARY_LOCATION}       if ( $] < 5.005 );

    # more or less taken from Moose' Makefile.PL
    if ( $params{CONFLICTS} )
    {
        my $ok = CheckConflicts(%params);
        exit(0) if ( $params{PREREQ_FATAL} and not $ok );
        my $cpan_smoker = grep { $_ =~ m/(?:CR_SMOKER|CPAN_REPORTER|AUTOMATED_TESTING)/ } keys %ENV;
        unless ( $cpan_smoker || $ENV{PERL_MM_USE_DEFAULT} )
        {
            sleep 4 unless ($ok);
        }
        delete $params{CONFLICTS};
    }

    WriteMakefile(%params);
}

sub _write_constants
{
    my @names = (
        qw(SG_ERROR_NONE SG_ERROR_INVALID_ARGUMENT SG_ERROR_ASPRINTF
          SG_ERROR_SPRINTF SG_ERROR_DEVICES SG_ERROR_DEVSTAT_GETDEVS
          SG_ERROR_DEVSTAT_SELECTDEVS SG_ERROR_DISKINFO SG_ERROR_ENOENT
          SG_ERROR_GETIFADDRS SG_ERROR_GETMNTINFO SG_ERROR_GETPAGESIZE
          SG_ERROR_HOST SG_ERROR_KSTAT_DATA_LOOKUP SG_ERROR_KSTAT_LOOKUP
          SG_ERROR_KSTAT_OPEN SG_ERROR_KSTAT_READ SG_ERROR_KVM_GETSWAPINFO
          SG_ERROR_KVM_OPENFILES SG_ERROR_MALLOC SG_ERROR_MEMSTATUS
          SG_ERROR_OPEN SG_ERROR_OPENDIR SG_ERROR_READDIR SG_ERROR_PARSE
          SG_ERROR_PDHADD SG_ERROR_PDHCOLLECT SG_ERROR_PDHOPEN SG_ERROR_PDHREAD
          SG_ERROR_PERMISSION SG_ERROR_PSTAT SG_ERROR_SETEGID SG_ERROR_SETEUID
          SG_ERROR_SETMNTENT SG_ERROR_SOCKET SG_ERROR_SWAPCTL SG_ERROR_SYSCONF
          SG_ERROR_SYSCTL SG_ERROR_SYSCTLBYNAME SG_ERROR_SYSCTLNAMETOMIB
          SG_ERROR_SYSINFO SG_ERROR_MACHCALL SG_ERROR_IOKIT SG_ERROR_UNAME
          SG_ERROR_UNSUPPORTED SG_ERROR_XSW_VER_MISMATCH SG_ERROR_GETMSG
          SG_ERROR_PUTMSG SG_ERROR_INITIALISATION SG_ERROR_MUTEX_LOCK
          SG_ERROR_MUTEX_UNLOCK),
        qw(sg_unknown_configuration sg_physical_host sg_virtual_machine
          sg_paravirtual_machine sg_hardware_virtualized),
        qw(sg_fs_unknown sg_fs_regular sg_fs_special sg_fs_loopback
          sg_fs_remote sg_fs_local sg_fs_alltypes),
        qw(SG_IFACE_DUPLEX_FULL SG_IFACE_DUPLEX_HALF SG_IFACE_DUPLEX_UNKNOWN),
        qw(SG_IFACE_DOWN SG_IFACE_UP),
        qw(SG_PROCESS_STATE_RUNNING SG_PROCESS_STATE_SLEEPING
          SG_PROCESS_STATE_STOPPED SG_PROCESS_STATE_ZOMBIE
          SG_PROCESS_STATE_UNKNOWN),
    );

    my $c_const  = File::Spec->catfile( dirname($xsfile), 'const-c.inc' );
    my $xs_const = File::Spec->catfile( dirname($xsfile), 'const-xs.inc' );

    # workaround #ifdef around each constant (stupid that it's not documented)
    open( my $c_fh, ">", $c_const ) or die("Cannot open $c_const for writing: $!");
    foreach my $name (@names)
    {
        print $c_fh "#define $name $name\n";
    }
    ExtUtils::Constant::WriteConstants(
        NAME         => 'Unix::Statgrab',
        NAMES        => \@names,
        DEFAULT_TYPE => 'UV',
        C_FH         => $c_fh,
        XS_FILE      => $xs_const,
    );
    close($c_fh);

    return;
}
